src/main/java/org/folio/service/transactions/batch/BatchTransactionHolder.java

```diff
@@ -0,0 +1,296 @@
+
+  private Future<Void> loadLinkedEncumbrances(DBConn conn) {
+    List<String> ids = allTransactionsToCreateOrUpdate.stream().map(tr -> {
+      if (List.of(PAYMENT, CREDIT).contains(tr.getTransactionType())) {
+        return tr.getPaymentEncumbranceId();
+      } else if (tr.getTransactionType() == PENDING_PAYMENT && tr.getAwaitingPayment() != null) {
+        return tr.getAwaitingPayment().getEncumbranceId();
+      } else {
+        return null;
+      }
+    }).filter(Objects::nonNull)
+      .distinct()
+      .toList();
+    return transactionDAO.getTransactionsByIds(ids, conn)
+      .map(transactions -> {
+        linkedEncumbrances = transactions;
+        return null;
+      });
+  }
+
+  private Future<Void> loadLinkedPendingPayments(DBConn conn) {
+    List<String> invoiceIds = allTransactionsToCreate.stream()
+      .filter(tr -> List.of(PAYMENT, CREDIT).contains(tr.getTransactionType()))
+      .map(Transaction::getSourceInvoiceId)
+      .distinct()
+      .toList();
+    if (invoiceIds.isEmpty()) {
+      return succeededFuture();
+    }
+    Criteria trTypeCrit = new Criteria()
+      .addField("'" + TRANSACTION_TYPE + "'")
+      .setOperation("=")
+      .setVal(PENDING_PAYMENT.value());
+    GroupedCriterias invoiceIdGroup = new GroupedCriterias();
+    invoiceIds.forEach(id -> invoiceIdGroup.addCriteria(
+      new Criteria()
+        .addField("'" + SOURCE_INVOICE_ID + "'")
+        .setOperation("=")
+        .setVal(id),
+      "OR"));
+    Criterion criterion = new Criterion();
+    criterion.addCriterion(trTypeCrit);
+    criterion.addGroupOfCriterias(invoiceIdGroup);
+    return transactionDAO.getTransactionsByCriterion(criterion, conn)
+      .map(transactions -> {
+        linkedPendingPayments = transactions;
+        return null;
+      });
+  }
+
+  private void setAllTransactions() {
+    allTransactions = Stream.of(allTransactionsToCreateOrUpdate.stream(), existingTransactions.stream(),
+        linkedEncumbrances.stream())
+      .flatMap(Function.identity())
+      .toList();
+  }
+
+  private Future<Void> loadFunds(DBConn conn) {
+    List<String> fundIds = allTransactions.stream()
+      .map(tr -> {
+        ArrayList<String> list = new ArrayList<>();
+        list.add(tr.getFromFundId());
+        list.add(tr.getToFundId());
+        return list;
+      })
+      .flatMap(Collection::stream)
+      .filter(Objects::nonNull)
+      .distinct()
+      .toList();
+    if (fundIds.isEmpty()) {
+      allFunds = emptyList();
+      return succeededFuture();
+    }
+    return fundService.getFundsByIds(fundIds, conn)
+      .map(funds -> {
+        allFunds = funds;
+        return null;
+      });
+  }
+
+  private Future<Void> loadBudgets(DBConn conn) {
+    Map<String, Set<String>> fiscalYearIdToFundIds = allTransactions.stream().collect(groupingBy(Transaction::getFiscalYearId,
+      flatMapping(tr -> (Stream<String>)Stream.of(tr.getFromFundId(), tr.getToFundId()).filter(Objects::nonNull), toSet())));
```

Comment:
(Stream<String>) - this cast is redudant

Reply:
I know, but it does not compile without it. A limitation of the Java compiler type inferring, I suppose.

Reply:
I actually can't make it fail anymore without the cast. Previously I could not make it compile without it. I will remove it.